File: python/sorting_algorithms/quick_sort_first.py
def partition(arr, start, end):
    pivot = arr[start]
    swap_index = start + 1
    print(f'arr: {arr}, pivot: {pivot}, start: {start}, swap_index: {swap_index}')

    for i in range(start + 1, end + 1):
        # print(pivot, i, arr[i] )
        # if arr[i] <= pivot:
        if pivot >= arr[i]:
            arr[i], arr[swap_index] = arr[swap_index], arr[i]
            # print(arr, i, swap_index)
            swap_index += 1

    arr[start], arr[swap_index - 1] = arr[swap_index - 1], arr[start]
    # print(arr, pivot, swap_index - 1)
    return swap_index - 1


def quick_sort(arr, start, end):
    if start < end:
        pivot_index = partition(arr, start, end)
        quick_sort(arr, start, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, end)
    return arr


arr1 = [26, 14, 78, 19, 99, 35, 26, 36, 37, 2]

# print(partition(arr1, 0, len(arr1) - 1))
print(quick_sort(arr1, 0, len(arr1) - 1))

File: javascript/sorting_algorithms/quickSortFirst.js
function partition(arr, start, end) {
    let pivot = arr[start];
    let swapIndex = start;

    const swap = (arr, indx1, indx2) => {
        [arr[indx1],arr[indx2]] = [arr[indx2],arr[indx1]];
    };

    for (let i = start + 1; i < arr.length; i++) {
        if (pivot > arr[i]) {
            swapIndex++;
            swap(arr, swapIndex, i);
            //swapIndex++;
        }
    }
    swap(arr, start, swapIndex);
    return swapIndex;
}

function quickSort(arr, start, end) {
    if (start < end) {
        let pivotIndex = partition(arr, start, end);
        quickSort(arr, start, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, end);
    }
    return arr;
}

//let arr1 = [26, 40, 41, 14, -18, 78, 19, 99, -10];
let arr1 = [0,1,2,3,4,5,6,7,8,9];

//let arr1 = [10,9,8,7,6,5,4,3,2,1,0];
//let arr1 = [9,0,-1];

console.log(quickSort(arr1, 0, arr1.length - 1));

